usr.bin/rs/rs.1
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)rs.1	8.2 (Berkeley) 12/30/93
.\" $FreeBSD$
.\"
.Dd April 7, 2015
.Dt RS 1
.Os
.Sh NAME
.Nm rs
.Nd reshape a data array
.Sh SYNOPSIS
.Nm
.Oo
.Fl Oo Cm csCS Oc Ns Op Ar x
.Oo Cm kKgGw Oc Ns Op Ar N
.Cm tTeEnyjhHmz
.Oc
.Op Ar rows Op Ar cols
.Sh DESCRIPTION
The
.Nm
utility reads the standard input, interpreting each line as a row
of blank-separated entries in an array,
transforms the array according to the options,
and writes it on the standard output.
With no arguments it transforms stream input into a columnar
format convenient for terminal viewing.
.Pp
The shape of the input array is deduced from the number of lines
and the number of columns on the first line.
If that shape is inconvenient, a more useful one might be
obtained by skipping some of the input with the
.Fl k
option.
Other options control interpretation of the input columns.
.Pp
The shape of the output array is influenced by the
.Ar rows
and
.Ar cols
specifications, which should be positive integers.
If only one of them is a positive integer,
.Nm
computes a value for the other which will accommodate
all of the data.
When necessary, missing data are supplied in a manner
specified by the options and surplus data are deleted.
There are options to control presentation of the output columns,
including transposition of the rows and columns.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl c Ns Ar x
Input columns are delimited by the single character
.Ar x .
A missing
.Ar x
is taken to be `^I'.
.It Fl s Ns Ar x
Like
.Fl c ,
but maximal strings of
.Ar x
are delimiters.
.It Fl C Ns Ar x
Output columns are delimited by the single character
.Ar x .
A missing
.Ar x
is taken to be `^I'.
.It Fl S Ns Ar x
Like
.Fl C ,
but padded strings of
.Ar x
are delimiters.
.It Fl t
Fill in the rows of the output array using the columns of the
input array, that is, transpose the input while honoring any
.Ar rows
and
.Ar cols
specifications.
.It Fl T
Print the pure transpose of the input, ignoring any
.Ar rows
or
.Ar cols
specification.
.It Fl k Ns Ar N
Ignore the first
.Ar N
lines of input.
.It Fl K Ns Ar N
Like
.Fl k ,
but print the ignored lines.
.It Fl g Ns Ar N
The gutter width (inter-column space), normally 2, is taken to be
.Ar N .
.It Fl G Ns Ar N
The gutter width has
.Ar N
percent of the maximum column width added to it.
.It Fl e
Consider each line of input as an array entry.
.It Fl n
On lines having fewer entries than the first line,
use null entries to pad out the line.
Normally, missing entries are taken from the next line of input.
.It Fl y
If there are too few entries to make up the output dimensions,
pad the output by recycling the input from the beginning.
Normally, the output is padded with blanks.
.It Fl h
Print the shape of the input array and do nothing else.
The shape is just the number of lines and the number of
entries on the first line.
.It Fl H
Like
.Fl h ,
but also print the length of each line.
.It Fl j
Right adjust entries within columns.
.It Fl w Ns Ar N
The width of the display, normally 80, is taken to be the positive
integer
.Ar N .
.It Fl m
Do not trim excess delimiters from the ends of the output array.
.It Fl z
Adapt column widths to fit the largest entries appearing in them.
.El
.Pp
With no arguments,
.Nm
transposes its input, and assumes one array entry per input line
unless the first non-ignored line is longer than the display width.
Option letters which take numerical arguments interpret a missing
number as zero unless otherwise indicated.
.Sh EXAMPLES
The
.Nm
utility can be used as a filter to convert the stream output
of certain programs (e.g.,
.Xr spell 1 ,
.Xr du 1 ,
.Xr file 1 ,
.Xr look 1 ,
.Xr nm 1 ,
.Xr who 1 ,
and
.Xr wc 1 )
into a convenient ``window'' format, as in
.Bd -literal -offset indent
% who | rs
.Ed
.Pp
This function has been incorporated into the
.Xr ls 1
program, though for most programs with similar output
.Nm
suffices.
.Pp
To convert stream input into vector output and back again, use
.Bd -literal -offset indent
% rs 1 0 | rs 0 1
.Ed
.Pp
A 10 by 10 array of random numbers from 1 to 100 and
its transpose can be generated with
.Bd -literal -offset indent
% jot \-r 100 | rs 10 10 | tee array | rs \-T > tarray
.Ed
.Pp
In the editor
.Xr vi 1 ,
a file consisting of a multi-line vector with 9 elements per line
can undergo insertions and deletions,
and then be neatly reshaped into 9 columns with
.Bd -literal -offset indent
:1,$!rs 0 9
.Ed
.Pp
Finally, to sort a database by the first line of each 4-line field, try
.Bd -literal -offset indent
% rs \-eC 0 4 | sort | rs \-c 0 1
.Ed
.Sh SEE ALSO
.Xr jot 1 ,
.Xr pr 1 ,
.Xr sort 1 ,
.Xr vi 1
.Sh HISTORY
The
.Nm
utility first appeared in
.Bx 4.2 .
.Sh AUTHORS
.An John A. Kunze
.Sh BUGS
.Bl -item
.It
Handles only two dimensional arrays.
.It
The algorithm currently reads the whole file into memory,
so files that do not fit in memory will not be reshaped.
.It
Fields cannot be defined yet on character positions.
.It
Re-ordering of columns is not yet possible.
.It
There are too many options.
.It
Multibyte characters are not recognized.
.It
Lines longer than
.Dv LINE_MAX
(2048) bytes are not processed and result in immediate termination of
.Nm .
.El


usr.bin/rs/rs.c
/*-
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static const char sccsid[] = "@(#)rs.c	8.1 (Berkeley) 6/6/93";
#endif /* not lint */

/*
 *	rs - reshape a data array
 *	Author:  John Kunze, Office of Comp. Affairs, UCB
 *		BEWARE: lots of unfinished edges
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <err.h>
#include <ctype.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static long	flags;
#define	TRANSPOSE	000001
#define	MTRANSPOSE	000002
#define	ONEPERLINE	000004
#define	ONEISEPONLY	000010
#define	ONEOSEPONLY	000020
#define	NOTRIMENDCOL	000040
#define	SQUEEZE		000100
#define	SHAPEONLY	000200
#define	DETAILSHAPE	000400
#define	RIGHTADJUST	001000
#define	NULLPAD		002000
#define	RECYCLE		004000
#define	SKIPPRINT	010000
#define	ICOLBOUNDS	020000
#define	OCOLBOUNDS	040000
#define ONEPERCHAR	0100000
#define NOARGS		0200000

static short	*colwidths;
static short	*cord;
static short	*icbd;
static short	*ocbd;
static int	nelem;
static char	**elem;
static char	**endelem;
static char	*curline;
static int	allocsize = BUFSIZ;
static int	curlen;
static int	irows, icols;
static int	orows = 0, ocols = 0;
static int	maxlen;
static int	skip;
static int	propgutter;
static char	isep = ' ', osep = ' ';
static char	blank[] = "";
static int	owidth = 80, gutter = 2;

static void	  getargs(int, char *[]);
static void	  getfile(void);
static int	  get_line(void);
static char	 *getlist(short **, char *);
static char	 *getnum(int *, char *, int);
static char	**getptrs(char **);
static void	  prepfile(void);
static void	  prints(char *, int);
static void	  putfile(void);
static void usage(void);

#define	INCR(ep) do {			\
	if (++ep >= endelem)		\
		ep = getptrs(ep);	\
} while(0)

int
main(int argc, char *argv[])
{
	getargs(argc, argv);
	getfile();
	if (flags & SHAPEONLY) {
		printf("%d %d\n", irows, icols);
		exit(0);
	}
	prepfile();
	putfile();
	exit(0);
}

static void
getfile(void)
{
	char *p;
	char *endp;
	char **ep;
	int c;
	int multisep = (flags & ONEISEPONLY ? 0 : 1);
	int nullpad = flags & NULLPAD;
	char **padto;

	while (skip--) {
		c = get_line();
		if (flags & SKIPPRINT)
			puts(curline);
		if (c == EOF)
			return;
	}
	get_line();
	if (flags & NOARGS && curlen < owidth)
		flags |= ONEPERLINE;
	if (flags & ONEPERLINE)
		icols = 1;
	else				/* count cols on first line */
		for (p = curline, endp = curline + curlen; p < endp; p++) {
			if (*p == isep && multisep)
				continue;
			icols++;
			while (*p && *p != isep)
				p++;
		}
	ep = getptrs(elem);
	do {
		if (flags & ONEPERLINE) {
			*ep = curline;
			INCR(ep);		/* prepare for next entry */
			if (maxlen < curlen)
				maxlen = curlen;
			irows++;
			continue;
		}
		for (p = curline, endp = curline + curlen; p < endp; p++) {
			if (*p == isep && multisep)
				continue;	/* eat up column separators */
			if (*p == isep)		/* must be an empty column */
				*ep = blank;
			else			/* store column entry */
				*ep = p;
			while (p < endp && *p != isep)
				p++;		/* find end of entry */
			*p = '\0';		/* mark end of entry */
			if (maxlen < p - *ep)	/* update maxlen */
				maxlen = p - *ep;
			INCR(ep);		/* prepare for next entry */
		}
		irows++;			/* update row count */
		if (nullpad) {			/* pad missing entries */
			padto = elem + irows * icols;
			while (ep < padto) {
				*ep = blank;
				INCR(ep);
			}
		}
	} while (get_line() != EOF);
	*ep = 0;				/* mark end of pointers */
	nelem = ep - elem;
}

static void
putfile(void)
{
	char **ep;
	int i, j, k;

	ep = elem;
	if (flags & TRANSPOSE)
		for (i = 0; i < orows; i++) {
			for (j = i; j < nelem; j += orows)
				prints(ep[j], (j - i) / orows);
			putchar('\n');
		}
	else
		for (i = k = 0; i < orows; i++) {
			for (j = 0; j < ocols; j++, k++)
				if (k < nelem)
					prints(ep[k], j);
			putchar('\n');
		}
}

static void
prints(char *s, int col)
{
	int n;
	char *p = s;

	while (*p)
		p++;
	n = (flags & ONEOSEPONLY ? 1 : colwidths[col] - (p - s));
	if (flags & RIGHTADJUST)
		while (n-- > 0)
			putchar(osep);
	for (p = s; *p; p++)
		putchar(*p);
	while (n-- > 0)
		putchar(osep);
}

static void
usage(void)
{
	fprintf(stderr,
		"usage: rs [-[csCS][x][kKgGw][N]tTeEnyjhHmz] [rows [cols]]\n");
	exit(1);
}

static void
prepfile(void)
{
	char **ep;
	int  i;
	int  j;
	char **lp;
	int colw;
	int max;
	int n;

	if (!nelem)
		exit(0);
	gutter += maxlen * propgutter / 100.0;
	colw = maxlen + gutter;
	if (flags & MTRANSPOSE) {
		orows = icols;
		ocols = irows;
	}
	else if (orows == 0 && ocols == 0) {	/* decide rows and cols */
		ocols = owidth / colw;
		if (ocols == 0) {
			warnx("display width %d is less than column width %d",
					owidth, colw);
			ocols = 1;
		}
		if (ocols > nelem)
			ocols = nelem;
		orows = nelem / ocols + (nelem % ocols ? 1 : 0);
	}
	else if (orows == 0)			/* decide on rows */
		orows = nelem / ocols + (nelem % ocols ? 1 : 0);
	else if (ocols == 0)			/* decide on cols */
		ocols = nelem / orows + (nelem % orows ? 1 : 0);
	lp = elem + orows * ocols;
	while (lp > endelem) {
		getptrs(elem + nelem);
		lp = elem + orows * ocols;
	}
	if (flags & RECYCLE) {
		for (ep = elem + nelem; ep < lp; ep++)
			*ep = *(ep - nelem);
		nelem = lp - elem;
	}
	if (!(colwidths = (short *) malloc(ocols * sizeof(short))))
		errx(1, "malloc");
	if (flags & SQUEEZE) {
		ep = elem;
		if (flags & TRANSPOSE)
			for (i = 0; i < ocols; i++) {
				max = 0;
				for (j = 0; *ep != NULL && j < orows; j++)
					if ((n = strlen(*ep++)) > max)
						max = n;
				colwidths[i] = max + gutter;
			}
		else
			for (i = 0; i < ocols; i++) {
				max = 0;
				for (j = i; j < nelem; j += ocols)
					if ((n = strlen(ep[j])) > max)
						max = n;
				colwidths[i] = max + gutter;
			}
	}
	/*	for (i = 0; i < orows; i++) {
			for (j = i; j < nelem; j += orows)
				prints(ep[j], (j - i) / orows);
			putchar('\n');
		}
	else
		for (i = 0; i < orows; i++) {
			for (j = 0; j < ocols; j++)
				prints(*ep++, j);
			putchar('\n');
		}*/
	else
		for (i = 0; i < ocols; i++)
			colwidths[i] = colw;
	if (!(flags & NOTRIMENDCOL)) {
		if (flags & RIGHTADJUST)
			colwidths[0] -= gutter;
		else
			colwidths[ocols - 1] = 0;
	}
	n = orows * ocols;
	if (n > nelem && (flags & RECYCLE))
		nelem = n;
	/*for (i = 0; i < ocols; i++)
		warnx("%d is colwidths, nelem %d", colwidths[i], nelem);*/
}

#define	BSIZE	(LINE_MAX * 2)
static char	ibuf[BSIZE];

static int
get_line(void)	/* get line; maintain curline, curlen; manage storage */
{
	static	int putlength;
	static	char *endblock = ibuf + BSIZE;
	char *p;
	int c, i;

	if (!irows) {
		curline = ibuf;
		putlength = flags & DETAILSHAPE;
	}
	else if (skip <= 0) {			/* don't waste storage */
		curline += curlen + 1;
		if (putlength) {	/* print length, recycle storage */
			printf(" %d line %d\n", curlen, irows);
			curline = ibuf;
		}
	}
	if (!putlength && endblock - curline < LINE_MAX + 1) { /* need storage */
		/*ww = endblock-curline; tt += ww;*/
		/*printf("#wasted %d total %d\n",ww,tt);*/
		if (!(curline = (char *) malloc(BSIZE)))
			errx(1, "file too large");
		endblock = curline + BSIZE;
		/*printf("#endb %d curline %d\n",endblock,curline);*/
	}
	for (p = curline, i = 0;; *p++ = c, i++) {
		if ((c = getchar()) == EOF)
			break;
		if (i >= LINE_MAX)
			errx(1, "maximum line length (%d) exceeded", LINE_MAX);
		if (c == '\n')
			break;
	}
	*p = '\0';
	curlen = i;
	return(c);
}

static char **
getptrs(char **sp)
{
	char **p;

	allocsize += allocsize;
	p = (char **)realloc(elem, allocsize * sizeof(char *));
	if (p == NULL)
		err(1, "no memory");

	sp += (p - elem);
	endelem = (elem = p) + allocsize;
	return(sp);
}

static void
getargs(int ac, char *av[])
{
	char *p;

	if (ac == 1) {
		flags |= NOARGS | TRANSPOSE;
	}
	while (--ac && **++av == '-')
		for (p = *av+1; *p; p++)
			switch (*p) {
			case 'T':
				flags |= MTRANSPOSE;
			case 't':
				flags |= TRANSPOSE;
				break;
			case 'c':		/* input col. separator */
				flags |= ONEISEPONLY;
			case 's':		/* one or more allowed */
				if (p[1])
					isep = *++p;
				else
					isep = '\t';	/* default is ^I */
				break;
			case 'C':
				flags |= ONEOSEPONLY;
			case 'S':
				if (p[1])
					osep = *++p;
				else
					osep = '\t';	/* default is ^I */
				break;
			case 'w':		/* window width, default 80 */
				p = getnum(&owidth, p, 0);
				if (owidth <= 0)
					errx(1, "width must be a positive integer");
				break;
			case 'K':			/* skip N lines */
				flags |= SKIPPRINT;
			case 'k':			/* skip, do not print */
				p = getnum(&skip, p, 0);
				if (!skip)
					skip = 1;
				break;
			case 'm':
				flags |= NOTRIMENDCOL;
				break;
			case 'g':		/* gutter space */
				p = getnum(&gutter, p, 0);
				break;
			case 'G':
				p = getnum(&propgutter, p, 0);
				break;
			case 'e':		/* each line is an entry */
				flags |= ONEPERLINE;
				break;
			case 'E':
				flags |= ONEPERCHAR;
				break;
			case 'j':			/* right adjust */
				flags |= RIGHTADJUST;
				break;
			case 'n':	/* null padding for missing values */
				flags |= NULLPAD;
				break;
			case 'y':
				flags |= RECYCLE;
				break;
			case 'H':			/* print shape only */
				flags |= DETAILSHAPE;
			case 'h':
				flags |= SHAPEONLY;
				break;
			case 'z':			/* squeeze col width */
				flags |= SQUEEZE;
				break;
			/*case 'p':
				ipagespace = atoi(++p);	(default is 1)
				break;*/
			case 'o':			/* col order */
				p = getlist(&cord, p);
				break;
			case 'b':
				flags |= ICOLBOUNDS;
				p = getlist(&icbd, p);
				break;
			case 'B':
				flags |= OCOLBOUNDS;
				p = getlist(&ocbd, p);
				break;
			default:
				usage();
			}
	/*if (!osep)
		osep = isep;*/
	switch (ac) {
	/*case 3:
		opages = atoi(av[2]);*/
	case 2:
		if ((ocols = atoi(av[1])) < 0)
			ocols = 0;
	case 1:
		if ((orows = atoi(av[0])) < 0)
			orows = 0;
	case 0:
		break;
	default:
		errx(1, "too many arguments");
	}
}

static char *
getlist(short **list, char *p)
{
	int count = 1;
	char *t;

	for (t = p + 1; *t; t++) {
		if (!isdigit((unsigned char)*t))
			errx(1,
	"option %.1s requires a list of unsigned numbers separated by commas", t);
		count++;
		while (*t && isdigit((unsigned char)*t))
			t++;
		if (*t != ',')
			break;
	}
	if (!(*list = (short *) malloc(count * sizeof(short))))
		errx(1, "no list space");
	count = 0;
	for (t = p + 1; *t; t++) {
		(*list)[count++] = atoi(t);
		printf("++ %d ", (*list)[count-1]);
		fflush(stdout);
		while (*t && isdigit((unsigned char)*t))
			t++;
		if (*t != ',')
			break;
	}
	(*list)[count] = 0;
	return(t - 1);
}

/*
 * num = number p points to; if (strict) complain
 * returns pointer to end of num
 */
static char *
getnum(int *num, char *p, int strict)
{
	char *t = p;

	if (!isdigit((unsigned char)*++t)) {
		if (strict || *t == '-' || *t == '+')
			errx(1, "option %.1s requires an unsigned integer", p);
		*num = 0;
		return(p);
	}
	*num = atoi(t);
	while (*++t)
		if (!isdigit((unsigned char)*t))
			break;
	return(--t);
}
